python_bindings/bindings.cpp

```diff
@@ -345,140 +348,153 @@ class Index {
+      new_index->ep_added=d["ep_added"].cast<bool>();
+      new_index->num_threads_default=d["num_threads"].cast<int>();
+      new_index->default_ef=d["ef"].cast<size_t>();
 
       if (index_inited_)
-        new_index->setAnnData(ann_params);
-
+        new_index->setAnnData(d);
 
       return new_index;
     }
 
     static Index<float> * createFromIndex(const Index<float> & index) {
-        /* WARNING:     Index::getIndexParams is not thread-safe with Index::addItems */
-        return createFromParams(index.getIndexParams()); 
+        return createFromParams(index.getIndexParams());
     }
 
-    
-    void setAnnData(const py::tuple t) {
-      /* WARNING: Index::setAnnData is not thread-safe with Index::addItems */
-      
+    void setAnnData(const py::dict d) { /* WARNING: Index::setAnnData is not thread-safe with Index::addItems */
+
+
       std::unique_lock <std::mutex> templock(appr_alg->global);
 
-      assert_true(appr_alg->offsetLevel0_ == t[0].cast<size_t>(), "Invalid value of offsetLevel0_ ");
-      assert_true(appr_alg->max_elements_ == t[1].cast<size_t>(), "Invalid value of max_elements_ ");
+      assert_true(appr_alg->offsetLevel0_ == d["offset_level0"].cast<size_t>(), "Invalid value of offsetLevel0_ ");
+      assert_true(appr_alg->max_elements_ == d["max_elements"].cast<size_t>(), "Invalid value of max_elements_ ");
+
+      appr_alg->cur_element_count = d["cur_element_count"].cast<size_t>();
 
-      appr_alg->cur_element_count = t[2].cast<size_t>();
+      assert_true(appr_alg->size_data_per_element_ == d["size_data_per_element"].cast<size_t>(), "Invalid value of size_data_per_element_ ");
+      assert_true(appr_alg->label_offset_ == d["label_offset"].cast<size_t>(), "Invalid value of label_offset_ ");
+      assert_true(appr_alg->offsetData_ == d["offset_data"].cast<size_t>(), "Invalid value of offsetData_ ");
 
-      assert_true(appr_alg->size_data_per_element_ == t[3].cast<size_t>(), "Invalid value of size_data_per_element_ ");
-      assert_true(appr_alg->label_offset_ == t[4].cast<size_t>(), "Invalid value of label_offset_ ");
-      assert_true(appr_alg->offsetData_ == t[5].cast<size_t>(), "Invalid value of offsetData_ ");
+      appr_alg->maxlevel_ = d["max_level"].cast<int>();
+      appr_alg->enterpoint_node_ = d["enterpoint_node"].cast<hnswlib::tableint>();
 
-      appr_alg->maxlevel_ = t[6].cast<int>();
-      appr_alg->enterpoint_node_ = t[7].cast<hnswlib::tableint>();
+      assert_true(appr_alg->maxM_ == d["max_M"].cast<size_t>(), "Invalid value of maxM_ ");
+      assert_true(appr_alg->maxM0_ == d["max_M0"].cast<size_t>(), "Invalid value of maxM0_ ");
+      assert_true(appr_alg->M_ == d["M"].cast<size_t>(), "Invalid value of M_ ");
+      assert_true(appr_alg->mult_ == d["mult"].cast<double>(), "Invalid value of mult_ ");
+      assert_true(appr_alg->ef_construction_ == d["ef_construction"].cast<size_t>(), "Invalid value of ef_construction_ ");
```

Comment:
I think the messages might confuse the user. I would add something like "Unpickle error:...".

Reply:
changed the error message to "Unpickle error"